### PR TITLE
feat: install provider if not installed when clicking on create new button (#2706)

### LINKED ISSUE #2706: Creating a podman machine is not straightforward for some workflows
### Is your enhancement related to a problem? Please describe

As a new user, I see the welcome page.
Then I can click on 'Go to resources' page

And from the preferences, I see resources page where we can create resources.

But from there, I should be able to install podman if it's not yet installed.

### Describe the solution you'd like

Option 1 could be to have the 'create new..' button being enabled but if Podman is not installed, Podman extension should ask the user if he wants to install and then proceed the creation of the podman machine at the end.

Also a text needs be displayed on the right if there are no container/kubernetes factory (like we did for OpenShift Local and DevSandbox extension)

### Describe alternatives you've considered

Less voted option is to have a 'install...' button instead of 'create new...' button but it will require from user more clicks to achieve a 'ready to use runtime for containers'

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by lstocchi**
### What does this PR do?

It checks if the provider is installed before proceeding to the `create new` page.

- if installed, it works as before.
- If not installed, it performs a check of the prerequisites, if everything is ok it asks the user to confirm the installation, if something is missing the user is asked to fix it

### Screenshot/screencast of this PR

If we can proceed with the installation
install-podman

When some requirement is missing
requirements-podman-new-2

### What issues does this PR fix or reference?

it resolves #2706 and https://github.com/containers/podman-desktop/issues/2553

### How to test this PR?

1. uninstall podman
2. go to the resources page and click on `create new` button

I have to write some tests but I wanted to hear from you if this workflow is ok. 
@mairin after talking with Florent this morning we agreed on proceeding with a dialog and then using this PR to discuss if there is a better way to drive this workflow. As a reminder, we started talking about it before devtools week https://github.com/containers/podman-desktop/issues/2706 .

### CODE REVIEW

**Comment by deboer-tim** [comment]
You already have a machine, so the workflow is a bit confusing. ;-) I'm assuming at the start there would be no machine, and a message in that space that says something along the lines of "Podman is not installed. Click Create New to ...".

The recording of the first case is confusing to me because it says "Let's install Podman" but it looks like there is a Next button which closes the dialog. (and then the dialog appears again?) If this dialog only appears when prerequisites are missing and you cannot proceed, it would help if the text was more clear and the only action was Close.

**Comment by lstocchi** [comment]
> You already have a machine, so the workflow is a bit confusing. ;-) I'm assuming at the start there would be no machine, and a message in that space that says something along the lines of "Podman is not installed. Click Create New to ...".

I was playing with installing/uninstalling podman so the machine is visible when it shouldn't. Different problem. But in a normal situation there should be no machine. I will record the video again to avoid confusion.
My understanding is that we should not write anything at first, just show the `Create New...` button as always and then show a different workflow if podman is not installed.
 
> The recording of the first case is confusing to me because it says "Let's install Podman" but it looks like there is a Next button which closes the dialog. (and then the dialog appears again?) If this dialog only appears when prerequisites are missing and you cannot proceed, it would help if the text was more clear and the only action was Close.

The idea is that we show all the prerequisites then the user should fix the missing ones and click Next to proceed. In my case i didn't install wsl so if i click next, it keeps showing me that something is wrong. 
I could also click on `create new` -> see what i miss -> reduce PD -> fix stuff -> reopen PD -> click next -> install podman

**Comment by mairin** [comment]
@lstocchi Hey! Overall I don't think this approach is bad, a few comments though:

### 1. Flow: If all requirements are met, do we show?

If the user meets all the requirements, do they have to see this dialog? (It's an extra unnecessary click then?) I'd suggest to hide dialog when all reqs are met.

### 2. Flow: showing requirements that are successfully met at all?

I think probably it's better to only show the reqs that aren't met. Because if the user has the 4 met requirements here, that's not actionable to them, it's just informational, and it takes up space that could make the dialog a little more overwhelming than it needs to be? It would be cleaner / more to-the-point to only display requirements that are not met so I'd suggest only showing unmet reqs here.

### 3. Can't find podman (when you have it?)

This might just be a symptom of the dev environment you're working in...! But I see you have a podman machine that is running:

Screenshot from 2023-06-12 12-00-14

Then, when the dialog it appears it says "We couldn't find Podman. Let's install it." But I could find it :) Since you already have a VM. Just pointing this out, bc it is a little weird to say you can't find it when there's already one running?
Screenshot from 2023-06-12 12-02-13

Might be better to just rename the dialog (if possible) to "Podman system requirements check", or for parallelism with the analogous mocked up onboarding screen, "Some system requirements are missing" and set the context it's for podman machine set up, and, this assumes we only display missing reqs (see #2 point above)

(Also sorry to include this, I see in my slow commenting it's already been posted as feedback! But I am lazy to renumber my points manually 😆 )

### 4. Color palette of the pre-req check dialog

The dialog has a background of charcoal-600 and the inset with the list of checks doesn't appear to be in the palette (it comes up as #52525b on my end, the closest might be charcoal-400 which is #4a4b4f)

I am assuming the dialog standard is to have a bg of charcoal-600 in which case I would suggest charcoal-800 for the inset that displays the list of checks. I suggest this, because it's a darker color than the dialog base background color (which is the pattern established in the onboarding wizard mockups), and has just enough contrast to stand out (charcoal-700 isn't quite enough contrast IMO.) 
 
### 5. Consistent language

I think we call these "system requirements" pretty consistently in the onboarding mockups so I would s/prerequisites/system requirements/g to be consistent (or we can change the onboarding design... main point is they should be consistent :) Better to have one consistent term to refer to them than have multiple terms referring to the same thing that could cause confusion.

### 6. Next button is clickable when reqs aren't met // how do I check if I met 'em?

... What happens if i press next and I haven't fixed my WSL2? 😬  Could it be greyed out until met? If I run the command, how do I get it to re-run the check? (does hitting next re-run the check?)

### 7.  Visual design of the dialog
 
 I would tweak the design just a little bit to be more parallel with some of the extension / onboarding setup designs... something like this (and this is mostly just aesthetics, integrating some of the suggestions i wrote above, and consistency with the onboarding mockups for the similar thing):
 
Screenshot from 2023-06-12 12-31-54

### 8. Flow: what is the next step after meeting the requirements?

Once you meet the requirements.... are you placed back in the settings > resources screen, or do you go straight to the install podman wizard/dialog?

**Comment by lstocchi** [comment]
@mairin Updated with your suggestions. Now i have this.
requirements-podman-new

> ### 6. Next button is clickable when reqs aren't met // how do I check if I met 'em?
> ... What happens if i press next and I haven't fixed my WSL2? 😬 Could it be greyed out until met? If I run the command, how do I get it to re-run the check? (does hitting next re-run the check?)

If you press Next without haven't fixed WSL2 it will show you the same exact view.

I don't think that it would be good to grey out the button as i don't see a way to listen to all the changes in the system. I should execute the check function every x milliseconds. I don't like it much TBH. 
Then most of the time i think the user should close/restart PD because the changes require to restart the machine.

Basically when the user clicks on the `Create New ...` button and the provider is not installed, there are 3 steps. 
-> check requirements -> install provider -> go to create new page
So clicking on Next makes the user proceeds.

> ### 8. Flow: what is the next step after meeting the requirements?
> Once you meet the requirements.... are you placed back in the settings > resources screen, or do you go straight to the install podman wizard/dialog?

podman wizard/dialog.
The gif below shows what happens if you meet all requirements. 
It is also the same behavior when you miss requirements, fix them and click next. 
You click on `Create new ...` -> you see the missing requirements dialog -> fix them -> click Next -> ask confirmation to install PD -> PD wizard/dialog -> installation complete -> `create new` page opened
install-podman

**Comment by benoitf** [comment]
out of context side note: how do you have two "OpenShift Local" entries in Resources page ?
do you have old id (before 1.0) and new id extension being installed ?

**Comment by lstocchi** [comment]
> out of context side note: how do you have two "OpenShift Local" entries in Resources page ? do you have old id (before 1.0) and new id extension being installed ?

yes

**Comment by mairin** [comment]
@lstocchi Wow, thanks - that's looking great! 

Tiny nit - the margin to the left of the failed check in the inlay seems maybe a little too wide, could probably be halved? (I know the mockup I did had more margin but maybe a little less would look more "natural")

The flow without reboot seems reasonable! If there is a reboot - does the user have to remember to reopen Podman Desktop? (Do we do autostart on Windows by default?)

**Comment by lstocchi** [comment]
> Tiny nit - the margin to the left of the failed check in the inlay seems maybe a little too wide, could probably be halved? (I know the mockup I did had more margin but maybe a little less would look more "natural")

Like this? Or a bit more/less?
image

> The flow without reboot seems reasonable! If there is a reboot - does the user have to remember to reopen Podman Desktop? (Do we do autostart on Windows by default?)

No, it gets autostarted by default if the user does not update the settings.

File: packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte [hunk 1]
```diff
@@ -360,4 +448,61 @@ async function startConnectionProvider(
...
+              <button
+                class="pf-c-button pf-m-primary"
+                type="button"
+                on:click="{() => doCreateNew(providerToBeInstalled.provider, providerToBeInstalled.displayName)}"
+                >Next</button>
+            </div>
+          </div>
+        </div>
+      </div>
+    </Modal>
```
Comment (benoitf) [inline packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte]: Hi, I think we started to move out Modal to specific .svelte component

so it benefits with a more dedicated scope/variables/functions only used by the modal, and testing is easier

Comment (lstocchi) [reply packages/renderer/src/lib/preferences/PreferencesResourcesRendering.svelte]: Done 👍

**Comment by benoitf** [comment]
working on my side (I changed some checks so I have errors)

I was just wondering why we have a next button that do nothing when checks are not correct ?
what should it do ?

https://github.com/containers/podman-desktop/assets/436777/ca511c04-69b9-4d90-9349-68e3fa1e3391

**Comment by lstocchi** [comment]
> working on my side (I changed some checks so I have errors)
> 
> I was just wondering why we have a next button that do nothing when checks are not correct ? what should it do ?

Seems to be a bug, when clicking next it should try to perform the install, so if you didn't fix the issue it will show the same exact page again

**Comment by lstocchi** [comment]
> working on my side (I changed some checks so I have errors)
> 
> I was just wondering why we have a next button that do nothing when checks are not correct ? what should it do ?

Do you have some error in the console? It works fine for me

**Comment by benoitf** [comment]
@lstocchi I'm not changing anything (for the test for example I changed the required version of macOS being an unknown one (more recent that was exists)

Maybe next button should be a `retry` button in that case or something different (like messsage stating that if I do required changes I could try again, etc) because it was not clear to me that it would try to redo the "pre-flight checks"

**Comment by lstocchi** [comment]
> @lstocchi I'm not changing anything (for the test for example I changed the required version of macOS being an unknown one (more recent that was exists)

I was asking about an error in the console because the dialog should close and re-open. In your case it seems doing nothing. And because the `displayModal` is the first thing that is updated when the `createNew` function is called, it's like that function is not actually called.

> Maybe next button should be a `retry` button in that case or something different (like messsage stating that if I do required changes I could try again, etc) because it was not clear to me that it would try to redo the "pre-flight checks"

Ok, I'll update the name

**Comment by lstocchi** [comment]
@benoitf 
requirements-podman-retry

**Comment by benoitf** [comment]
thanks, I think issue on my side was also that the check was so fast that I was under the impression that the dialog was never disappearing/reappearing

**Review by benoitf — APPROVED** [review]
tested and works for me 🎉

**Pull request merged**